BurstBuffer/FsBuffer implementation

jrmadsen opened the pull request:
- New FsBuffer (file system buffer) class in cache.py to handle burst-buffe/file-system caching of cache objects
- Disable with Cache.use_fsbuffer = False (python)
- Disable with TOAST_NO_FSBUFFER > 0 (env variable)
- Set fscache temporary directory with FsCache.temporary_directory (python)
- Set fscache temporary directory with TOAST_TMPDIR (env variable)
- Added BurstBuffer support to SLURM examples
- Added TOAST_TMPDIR to all SLURM and shell examples

jrmadsen committed:
BurstBuffer/FsBuffer implementation

- New FsBuffer (file system buffer) class in cache.py to handle burst-buffe/file-system caching of cache objects
- Disable with Cache.use_fsbuffer = False (python)
- Disable with TOAST_NO_FSBUFFER > 0 (env variable)
- Set fscache temporary directory with FsCache.temporary_directory (python)
- Set fscache temporary directory with TOAST_TMPDIR (env variable)
- Added BurstBuffer support to SLURM examples
- Added TOAST_TMPDIR to all SLURM and shell examples

keskitalo commented on src/python/cache.py:
```diff
@@ -308,9 +452,35 @@ def reference(self, name):
         return ref
 
 
+    def clone(self, name):
```
I don't think this is what we had in mind during the discussion. We wanted a `copy` method that would return a plain Numpy ndarray. Operators using the ndarrays could then use the acquired copies as they like and have them garbage collected once they are done.

keskitalo reviewed (commented):
This is a great start but it is missing the `swap` and `copy` methods we agreed upon. Unswapping would then happen implicitly when the client code asks for a reference to a swapped object. Asking for a copy, instead, would not wipe the swapped object but rather the swapped object would be loaded and copied without anything left in the cache memory.

jrmadsen reviewed (commented):
`FsBuffer` provides the capabilities to read/write to the file-system/burst-buffer. Any of the additional edits involving the logic in `Cache` or method names in `FsBuffer` can be replaced/modified to y'all's preference.

I think the `swap` and `copy` methods exist, just under different names. I avoided using the word _swap_ as a function because that function name is possibly confused with function switching the assignment of variables (more easily expressed in C):

```c
void swap(int* a, int* b)
{
    int* tmp = b;
    b = a;
    a = tmp;
}

struct simple
{
    void swap(int* rhs) { swap(m_data, rhs); }
private:
    int* m_data;
}
```

A `swap` method in the above interpretation doesn't seem to make much sense though when dealing with file-systems. Instead, one would just have FsBuffer objects for `a` and `b` objects.  

Thus, instead of `swap` and `copy` as the names, I used `unload` and `load` in `FsBuffer`. Loading a numpy array from the file-system is inherently an operation that returns a copy of the data since this is read-only memory. And I called writing a copy of the numpy array in RAM to physical swap "unloading". Maybe `store` would be a better function name?

Cache is then responsible for handling freeing the RAM. For this, I provided `Cache.free(...)` as an alternative to `Cache.clear(...)`. Where the clear function will remove all references in RAM and FsBuffer references, the free function will remove all the references in RAM but keep the FsBuffer references so that when a reference is requested later, it can load the reference from the FsBuffer. The `Cache.clone(...)` method allows one to obtain a copy of the data in the way that you requested: if there is a FsBuffer reference, if will load the copy and not insert a reference into the Cache object. I think I should switch the logic though, so that it checks for the RAM copy first and then use the FsBuffer as the fallback instead of vice versa.

The logic that I provided (that you should replace if you don't like it) for handling whether to implicitly or explicitly remove the Cache reference is handled through the `remove_fsbuffer=...` arguments. By default, nothing is implicitly removed from RAM. The current default arguments to previously existing methods do not implicitly free any memory, as requested in our discussion.

jrmadsen committed:
CMake updates

jrmadsen committed:
Merge branch 'master' into burst-buffer

jrmadsen committed:
FsBuffer updates + auto_disk_array

- auto_disk_array is a wrapper around numpy that when all references are garbage collected, it places the array in an FsBuffer object
- Cache has two new dictionaries: _fsbuffers (storing FsBuffer objects) and _auto_refs (dictionary counting the number of auto_disk_array objects that exist)
- Many unittests testing and verifying the behavior of these utilities

jrmadsen committed:
Merge remote-tracking branch 'origin/master' into burst-buffer

jrmadsen committed:
Minor tests/cache.py updates + documentation for filesystem-cache

jrmadsen committed:
Merge remote-tracking branch 'origin/master' into burst-buffer

jrmadsen committed:
Fix to travis (download TiMemory) + others

- Removed accidental commit of local FindMPI.cmake
- {in,out}_no fix for auto_disk_array.__array_ufunc___

jrmadsen committed:
Merge branch 'master' into burst-buffer

jrmadsen committed:
Merge remote-tracking branch 'origin/master' into burst-buffer

jrmadsen committed:
Minor cache.py updates + alpha of Travis CDash reporting

jrmadsen committed:
Disabled USE_AATM in Travis CMake

jrmadsen committed:
Minor Travis + Testing.cmake updates

jrmadsen committed:
Travis: CTest + CDash changes

jrmadsen committed:
Burst-buffer passing all unit-tests

- tod/tod.py must initialize Cache via Cache(use_fscache=False) for unknown reasons
- other than tod/tod.py, setting env var TOAST_USE_FSCACHE=1 (always use file-system cache) works
- cache.py still have some "verbose"/debug stuff, however this is only printed with env var TOAST_CACHE_VERBOSE=1
- Reduced max processors to 8 on Travis
- Some of the travis tests run with TOAST_USE_FSCACHE=1

jrmadsen committed:
Merge remote-tracking branch 'origin/master' into burst-buffer

jrmadsen committed:
Fully working burst-buffer

- Burst-buffer fully working after src/python/tod/tod.py issue resolved. However, the format needed to be switched:
    1) ref = self.cache.reference(self._pos)[start:start+n, :]
        ref[:, :] = pos
   2) self.cache.reference(self._pos)[start:start+n, :] = pos
where #1 must be replaced with #2 style
- Cleaned up cache.py
- All unit tests passing
- Added running example shell scripts to Travis
- src/python/tests/runner.py has a timemory decorator

jrmadsen committed:
Reduced verbosity in Travis when TOAST_USE_FSCACHE enabled (causing error)

jrmadsen committed:
Fix for PySM unit test

- Had to set pixels cache in OpPointingHpix to never use file-system cache due to issue with toast healpix_pointing_matrix failing @ ct.pointing_healpix_matrix(hpix. hpix, nest, eps, cal, mode, pdata, hwpang, flags, pixels, weights)

jrmadsen committed:
Fix to SLURM examples when BB not available

- Addresses failing tests for small_satellite on Edison and Cori-Haswell

jrmadsen committed:
Removed forcing CMAKE_INSTALL_PYTHONDIR

- this messes up the Docker installation to /usr

jrmadsen committed:
Optional docker/shifter support for SLURM examples

- Define SHIFTER_SBATCH=SBATCH, SHIFTER_IMAGE=<image>, and SHIFTER=shifter to enable shifter when running ./generate_slurm directly
- From CMake, just define SHIFTER_IMAGE to enable shifter (default = OFF)
